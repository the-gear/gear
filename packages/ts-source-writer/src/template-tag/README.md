# Jak to muze fungovat

- `ts` template tag vytvori fragment â€” pouze neco jako AST.

- Kazdy vystup z `ts` (`TsSource`) ma metodu `toString`:
```ts
toString() {
  return new SourceResolver(this).toString();
}
```

to znamena, ze v jakekoliv fazi lze projit cely kod sestaveny
z `ts` template tagu. SourceResolver se stara o setaveni vseho duleziteho,
mozna pomoci WeakMap a WeakSet
