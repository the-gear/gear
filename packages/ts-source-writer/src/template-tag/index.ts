export * from './value';
export * from './identifier';
export * from './index';
export * from './source-resolver';
export * from './template-tag';
export * from './ts-source';
export * from './utils';
