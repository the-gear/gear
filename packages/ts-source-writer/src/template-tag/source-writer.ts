import { TsSource, isTsSource } from './ts-source';
import { isPrimitiveValue, isWithKeys } from './utils';

class Ref {
  parent: Ref | null;
  count = 0;
  isRecursive: boolean = false;
  suggestedNames = new Map<string, number>();

  constructor(parent: Ref | null) {
    this.parent = parent;
  }

  inc(suggestedNames?: string[]) {
    this.count++;
    if (suggestedNames) {
      suggestedNames.forEach((name) => this.suggestName(name));
    }
  }

  suggestName(name: string) {
    this.suggestedNames.set(name, (this.suggestedNames.get(name) || 0) + 1);
  }
}

export class SourceWriter {
  private source: string[] = [];
  private refs = new Map<unknown, Ref>();

  resolve(source: TsSource): this {
    source.resolve && source.resolve(this);
    return this;
  }

  addRef(data: unknown, suggestedNames?: string[]): this {
    this.addRefRecursive(data, suggestedNames || [], null, new Set());
    return this;
  }

  write(strOrTsSource: string | TsSource): this {
    this.writeRefs(this.source);
    if (typeof strOrTsSource === 'string') {
      this.source.push(strOrTsSource);
    } else if (isTsSource(strOrTsSource) && strOrTsSource.write) {
      strOrTsSource.write(this);
    } else {
      throw new Error('Invalid value');
    }
    return this;
  }

  getTsCode() {
    return this.source.join('');
  }

  private addRefRecursive(
    data: unknown,
    suggestedNames: string[],
    parent: Ref | null,
    seen: Set<unknown>,
  ) {
    let ref = this.refs.get(data);
    if (ref && seen.has(data)) {
      ref.isRecursive = true;
      return;
    }
    if (!ref) {
      if (isPrimitiveValue(data)) return;
      ref = new Ref(parent);
      this.refs.set(data, ref);
    }
    seen.add(data);
    ref.inc(suggestedNames);

    if (isWithKeys(data)) {
      for (const [key, value] of Object.entries(data)) {
        const safeKeyId = key;
        const newSuggestedNames = [
          safeKeyId,
          ...suggestedNames.map((name) => `${name}$${safeKeyId}`),
        ];
        this.addRefRecursive(value, newSuggestedNames, ref, seen);
      }
    }
    seen.delete(data);
  }

  writeRef(
    data: unknown,
    source: string[] = this.source,
    ref: Ref | undefined = this.refs.get(data),
  ): this {
    if (!ref) {
      source.push(`/* ??? ${typeof data} */`);
      return this;
    }
    source.push(
      `/* c:${ref.count} [${[...ref.suggestedNames.entries()]
        .map(([k, v]) => `${k}: ${v}`)
        .join(', ')}] */`,
    );
    return this;
  }

  private writeRefs(source: string[] = this.source) {
    for (const [data, ref] of this.refs.entries()) {
      if (ref.count > 0 || ref.isRecursive) {
        this.writeRef(data, source, ref);
      }
    }
  }
}
