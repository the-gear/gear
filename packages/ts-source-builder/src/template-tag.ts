import { RawSource } from './raw-source';
import { SourceCollection } from './source-collection';
import { TsSource, isTsSource } from './source-atom';
import { Identifier } from './identifier';
import { ImportedIdentifier } from './identifier-import';

function getDedentStr(str: string): string | null {
  const match = str.match(/(?:\r\n|\r|\n)(\s*)$/);
  return (match && match[1]) || '';
}

function dedent(str: string, dedentStr: string | null): string {
  if (!dedentStr) return str;
  return str
    .split(/\r\n|\r|\n/)
    .map((line) => (line.startsWith(dedentStr) ? line.slice(dedentStr.length) : line))
    .join('\n');
}

export function ts(
  literals: TemplateStringsArray,
  ...placeholders: Array<TsSource | string | number | boolean | null>
): SourceCollection {
  const placeholdersLength = placeholders.length;
  const literalsLength = literals.length;
  if (literalsLength - 1 !== placeholdersLength) {
    throw new Error(
      `Length mismatch: (${literalsLength}) literalsLength - 1 !== placeholdersLength (${placeholdersLength})`,
    );
  }

  let lastLiteral = literals[placeholdersLength];
  const dedentStr = getDedentStr(lastLiteral);
  lastLiteral = dedent(lastLiteral, dedentStr);

  const fragments: TsSource[] = [];
  for (let i = 0; i < placeholdersLength; i++) {
    const literal = dedent(literals[i], dedentStr);
    fragments.push(new RawSource(literal));
    const source = placeholders[i];
    if (isTsSource(source)) {
      fragments.push(source);
    } else if (source === null) {
      fragments.push(new RawSource('null'));
    } else {
      const typeofSource = typeof source;
      switch (typeofSource) {
        case 'string':
        case 'number':
        case 'boolean': {
          fragments.push(new RawSource(JSON.stringify(source)));
          break;
        }
        default: {
          throw new Error(
            `[ts-source-builder] Invalid value of type ${typeofSource}, placeholder ${i +
              1}: ${JSON.stringify(source)}, after code block ${literal}`,
          );
        }
      }
    }
  }

  // placeholdersLength is equal to literalsLength - 1
  fragments.push(new RawSource(lastLiteral));

  return new SourceCollection(fragments);
}

export function value(val: any): RawSource {
  return new RawSource(JSON.stringify(val));
}

export function identifier(name?: string, noRename?: boolean): Identifier {
  return new Identifier({ name, noRename });
}

export function importId(
  from: string,
  name?: string | null,
  alias?: string,
  noRename?: boolean,
): ImportedIdentifier {
  return new ImportedIdentifier({
    from,
    name,
    alias,
    noRename,
  });
}

// expose for easy of use
ts.val = value;
ts.id = identifier;
ts.import = importId;
