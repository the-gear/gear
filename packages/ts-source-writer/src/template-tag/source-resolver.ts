import { Source, isSource } from './source';
import { isPrimitiveValue, isWithKeys, getPropertyName, isSafeName } from './utils';

enum RefStage {
  WILL_WRITE,
  WILL_WRITE_DEP,
  WRITTING,
  WRITTEN,
}

class Ref {
  seen: number = 1;
  parents = new Set<Ref>();
  suggestedNames = new Map<string, number>();
  isRecursive: boolean = false;
  identifier: string | null = null;
  isExport: boolean = false;
  stage: RefStage = RefStage.WILL_WRITE;

  constructor(public data: unknown) {}

  addParent(parent: Ref) {
    this.parents.add(parent);
    return this;
  }

  suggestNames(suggestedNames: string[]) {
    if (suggestedNames) {
      suggestedNames.forEach((name) => this.suggestName(name));
    }
  }

  suggestName(name: string) {
    this.suggestedNames.set(name, (this.suggestedNames.get(name) || 0) + 1);
  }

  shouldWriteDep(): boolean {
    return this.parents.size > 1;
  }
}

export class SourceResolver {
  private source: string[] = [];
  private sourceDefs: string[] = [];
  private refs = new Map<unknown, Ref>();
  private suggestedRefNames = new Map<string, Set<Ref>>();
  private identifiers = new Set<string>();
  private resolvedSources = new Set<Source>();

  resolve(source: Source): this {
    if (source.resolve && !this.resolvedSources.has(source)) {
      this.resolvedSources.add(source);
      source.resolve(this);
    }
    return this;
  }

  write(anything: unknown): this {
    this.writeRefDeps();

    if (isSource(anything)) {
      anything.write && anything.write(this);
    } else {
      this.writeValue(anything);
    }
    return this;
  }

  addRef(data: unknown, suggestedNames?: string[]): this {
    this.addRefRecursive(data, suggestedNames || [], null, new Set());
    return this;
  }

  /**
   * Try reserve identifier. Return true if identifier can be used
   */
  tryIdentifier(name: string): boolean {
    if (this.identifiers.has(name)) return false;
    this.identifiers.add(name);
    return true;
  }

  reserveIdentifier(name: string): void {
    if (!this.tryIdentifier(name)) throw new Error(`Identifier ${name} is already taken`);
  }

  private suggestRefName(name: string, ref: Ref) {
    let refSet = this.suggestedRefNames.get(name);
    if (!refSet) {
      refSet = new Set();
      this.suggestedRefNames.set(name, refSet);
    }
    refSet.add(ref);
  }

  private addRefRecursive(
    data: unknown,
    suggestedNames: string[],
    parent: Ref | null,
    seen: Set<unknown>,
  ) {
    let ref = this.refs.get(data);
    if (ref) {
      ref.seen++;
      if (seen.has(data)) {
        ref.isRecursive = true;
        return;
      }
    }
    if (!ref) {
      if (isPrimitiveValue(data)) return;
      ref = new Ref(data);
      this.refs.set(data, ref);
    }
    seen.add(data);
    if (parent) ref.addParent(parent);
    ref.suggestNames(suggestedNames);
    suggestedNames.forEach((name) => this.suggestRefName(name, ref as Ref));

    if (isWithKeys(data)) {
      for (const [key, value] of Object.entries(data)) {
        const safeKeyId = key;
        const newSuggestedNames = [
          safeKeyId,
          ...suggestedNames.map((name) => `${name}$${safeKeyId}`),
        ];
        this.addRefRecursive(value, newSuggestedNames, ref, seen);
      }
    }
    seen.delete(data);
  }

  getFreeIdentifier(prefix: string = '$'): string {
    let i: number = 0;
    let id: string;
    do {
      id = `${prefix}$${(++i).toString(36)}`;
    } while (!this.tryIdentifier(id));
    return id;
  }

  getIdentifierFor(ref: Ref) {
    if (ref.identifier) {
      return ref.identifier;
    }
    const names = [...ref.suggestedNames.entries()];
    // sort by suggestion count
    names.sort((a, b) => b[1] - a[1]);
    let name: string;
    for (let i = 0; i < names.length; i++) {
      name = names[i][0];
      const refSet = this.suggestedRefNames.get(name);
      if (!refSet) {
        throw new Error(`RefSet not contain entry for '${name}'`);
      }
      if (refSet.size === 1 && refSet.has(ref)) {
        if (this.tryIdentifier(name)) {
          ref.identifier = name;
          return name;
        }
      }
    }

    return this.getFreeIdentifier('$' + typeof ref.data);
  }

  private captureWrite(cb: () => void): string {
    const backup = this.source;
    this.source = [];
    cb();
    const result = this.source.join('');
    this.source = backup;
    return result;
  }

  private writeRefDeps(refs = this.refs.values()) {
    for (const ref of refs) {
      if (ref && ref.shouldWriteDep() && !ref.identifier) {
        this.writeRefDep(ref);
      }
    }
  }

  private writeRefDep(ref: Ref) {
    if (ref.stage !== RefStage.WILL_WRITE) return;
    const id = this.getIdentifierFor(ref);
    this.sourceDefs.push(
      this.captureWrite(() => {
        this.writeCode(`${ref.isExport ? 'export ' : ''}const ${id} = `);
        ref.stage = RefStage.WILL_WRITE_DEP;
        this.writeValue(ref.data);
        this.writeCode(';\n');
      }),
    );
  }

  writeObject(obj: {}): this {
    const ref = this.refs.get(obj);
    if (!ref) {
      throw new Error('[writeObject] Ref not found');
    }

    if (ref.stage === RefStage.WILL_WRITE && ref.shouldWriteDep()) {
      this.writeRefDep(ref);
    }

    if (ref.stage === RefStage.WRITTEN) {
      if (!ref.identifier) {
        throw new Error('Object is written but have no identifier: ' + JSON.stringify(obj));
      }
      this.writeCode(ref.identifier);
      return this;
    }

    if (ref.stage === RefStage.WRITTING) {
      throw new Error('Unhandled recursion');
    }

    if (ref.stage === RefStage.WILL_WRITE || ref.stage === RefStage.WILL_WRITE_DEP) {
      ref.stage = RefStage.WRITTING;
      this.writeCode('({');
      let sep = '';
      for (const [key, value] of Object.entries(obj)) {
        this.writeCode(sep);
        const propName = getPropertyName(key);
        this.writeCode(propName);
        const valueCode = this.captureWrite(() => {
          this.writeValue(value);
        });
        if (valueCode !== propName || !isSafeName(propName)) {
          this.writeCode(': ');
          this.writeCode(valueCode);
        }
        sep = ', ';
      }
      this.writeCode('})');
      ref.stage = RefStage.WRITTEN;
      return this;
    }

    throw new Error(`[writeObject] Unhandled case '${ref.identifier}' ${RefStage[ref.stage]}`);
  }

  writeArray(array: unknown[]): this {
    const ref = this.refs.get(array);
    if (!ref) {
      throw new Error('[writeArray] Ref not found');
    }

    if (ref.stage === RefStage.WILL_WRITE && ref.shouldWriteDep()) {
      this.writeRefDep(ref);
    }

    if (ref.stage === RefStage.WRITTEN) {
      if (!ref.identifier) {
        throw new Error('Array is written but have no identifier');
      }
      this.writeCode(ref.identifier);
      return this;
    }

    if (ref.stage === RefStage.WRITTING) {
      throw new Error('Unhandled recursion');
    }

    if (ref.stage === RefStage.WILL_WRITE || ref.stage === RefStage.WILL_WRITE_DEP) {
      ref.stage = RefStage.WRITTING;
      this.writeCode('[');
      let sep = '';
      for (let i = 0; i < array.length; i++) {
        const value = array[i];
        this.writeCode(sep);
        this.writeValue(value);
        sep = ', ';
      }
      this.writeCode(']');
      ref.stage = RefStage.WRITTEN;
      return this;
    }

    throw new Error(`[writeObject] Unhandled case '${ref.identifier}' ${RefStage[ref.stage]}`);
  }

  writeFunction(fn: Function): this {
    // TODO
    return this.writeCode(fn.toString());
  }

  writeValue(val: unknown): this {
    switch (typeof val) {
      case 'string':
        return this.writeString(val);
      case 'number':
        return this.writeNumber(val);
      case 'boolean':
        return this.writeBoolean(val);
      case 'bigint':
        return this.writeBigInt(val);
      case 'undefined':
        return this.writeUndefined();
      case 'object': {
        if (val === null) return this.writeNull();
        if (Array.isArray(val)) return this.writeArray(val);
        return this.writeObject(val);
      }
      case 'symbol':
        return this.writeSymbol(val);
      case 'function':
        return this.writeFunction(val);
    }
    throw new Error(`[writeValue]: does not know how to write`);
  }

  getTsCode(): string {
    const result: string[] = [];
    if (this.sourceDefs.length) {
      result.push('// #region data definitions');
      result.push(this.sourceDefs.join('\n').trim());
      result.push('// #endregion data definitions\n');
    }
    result.push(this.source.join(''));
    return result.join('\n');
  }

  writeCode(code: string): this {
    this.source.push(code);
    return this;
  }

  writeBigInt(bigint: bigint): this {
    return this.writeCode(`BigInt('${bigint.toString()}')`);
  }

  writeBoolean(bool: boolean): this {
    return this.writeCode(bool ? 'true' : 'false');
  }

  writeNumber(n: number): this {
    if (Number.isNaN(n)) return this.writeCode('Number.NaN');
    if (n === Number.POSITIVE_INFINITY) return this.writeCode('Number.POSITIVE_INFINITY');
    if (n === Number.NEGATIVE_INFINITY) return this.writeCode('Number.NEGATIVE_INFINITY');
    return this.writeCode(n.toString());
  }

  writeString(str: string): this {
    return this.writeCode(JSON.stringify(str));
  }

  writeSymbol(_sym: symbol): this {
    // Well, I can register symbols on module and emit reference here, but...
    // return this.writeCode(`undefined /* ${sym.toString().replace(/\*\//g, '* /')} */`);
    throw new Error(`TODO: Symbol`);
  }

  writeUndefined() {
    return this.writeCode('undefined');
  }

  writeNull(): this {
    return this.writeCode('null');
  }
}
