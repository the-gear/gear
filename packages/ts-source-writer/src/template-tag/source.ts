import { SourceResolver } from './source-resolver';

export const __Source = Symbol();

export function isSource(value: any): value is Source {
  return value && value[__Source];
}

export function assertSource(value: Source): void;
export function assertSource(value: any): never;
export function assertSource(value: any) {
  if (!isSource(value)) throw new Error('Assertion failed');
}

export interface Source {
  readonly [__Source]: true;
  readonly dependencies?: ReadonlyArray<Source>;
  resolve?: (resolver: SourceResolver) => void;
  write?: (resolver: SourceResolver) => void;
}

export class SourceFragment implements Source {
  readonly [__Source] = true;

  toString(): string {
    return new SourceResolver()
      .resolve(this)
      .write(this)
      .getTsCode();
  }
}

export class SourceFragments extends SourceFragment {
  readonly dependencies: ReadonlyArray<Source>;

  constructor(fragments: ReadonlyArray<Source>) {
    super();
    this.dependencies = fragments;
  }

  resolve(resolver: SourceResolver) {
    this.dependencies.forEach((dep) => dep.resolve && dep.resolve(resolver));
    return this;
  }

  write(resolver: SourceResolver) {
    this.dependencies.forEach((dep) => dep.write && dep.write(resolver));
  }
}

export class RawSource extends SourceFragment {
  constructor(private source: string) {
    super();
  }

  write(resolver: SourceResolver) {
    resolver.writeCode(this.source);
  }

  toString(): string {
    return this.source;
  }
}
