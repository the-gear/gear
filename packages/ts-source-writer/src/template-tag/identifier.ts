import { SourceFragment } from './source';
import { SourceResolver } from './source-resolver';

export type IdentifierConfig = {
  name?: string;
  possibleNames?: string[];
  noRename?: boolean;
  isExported?: boolean;
};

export type IdentifierOptions = IdentifierConfig | string | string[];

export class Identifier extends SourceFragment implements IdentifierConfig {
  name?: string;
  possibleNames?: string[];
  noRename?: boolean;
  isExported?: boolean;
  resolvedName?: string;

  constructor(config?: IdentifierOptions) {
    super();
    if (typeof config === 'string') {
      this.name = config;
      this.noRename = true;
    } else if (typeof config !== 'undefined') {
      if (Array.isArray(config)) {
        this.possibleNames = config;
      } else {
        this.name = config.name;
        this.noRename = config.noRename;
      }
    }
  }

  resolve(resolver: SourceResolver) {
    if (this.name) {
      resolver.reserveIdentifier(this.name);
      this.resolvedName = this.name;
    } else if (this.possibleNames && this.possibleNames.length) {
      for (const name of this.possibleNames) {
        if (resolver.tryIdentifier(name)) {
          this.resolvedName = name;
          return;
        }
      }
      this.resolvedName = resolver.getFreeIdentifier(this.possibleNames[0]);
    } else {
      this.resolvedName = resolver.getFreeIdentifier('$id');
    }
  }

  write(resolver: SourceResolver) {
    resolver.writeCode(this.resolvedName as string);
  }
}
