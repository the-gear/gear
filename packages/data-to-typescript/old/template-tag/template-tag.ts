import { isSource, RawSource, SourceFragments, Source } from './source';
import { IdentifierOptions, Identifier } from './identifier';
import { value } from './value';

function getDedentStr(str: string): string | null {
  const match = str.match(/(?:\r\n|\r|\n)(\s*)$/);
  return (match && match[1]) || '';
}

function dedent(str: string, dedentStr: string | null): string {
  if (!dedentStr) return str;
  return str
    .split(/\r\n|\r|\n/)
    .map((line) => (line.startsWith(dedentStr) ? line.slice(dedentStr.length) : line))
    .join('\n');
}

export function ts(
  literals: TemplateStringsArray,
  ...placeholders: Array<Source | null>
): SourceFragments {
  const placeholdersLength = placeholders.length;
  const literalsLength = literals.length;
  if (literalsLength - 1 !== placeholdersLength) {
    throw new Error(
      `Length mismatch: (${literalsLength}) literalsLength - 1 !== placeholdersLength (${placeholdersLength})`,
    );
  }

  let lastLiteral = literals[placeholdersLength];
  const dedentStr = getDedentStr(lastLiteral);
  lastLiteral = dedent(lastLiteral, dedentStr);

  const fragments: Source[] = [];
  for (let i = 0; i < placeholdersLength; i++) {
    const literal = dedent(literals[i], dedentStr);
    fragments.push(new RawSource(literal));
    const source = placeholders[i];
    if (isSource(source)) {
      fragments.push(source);
    } else if (source === null || source === undefined) {
      // ignore, this allow conditionals
    } else {
      throw new Error(
        `[ts-source-writer] Invalid (unescaped) value, placeholder ${i + 1}: ${JSON.stringify(
          source,
        )}, after code block ${literal}`,
      );
    }
  }

  // placeholdersLength is equal to literalsLength - 1
  fragments.push(new RawSource(lastLiteral));

  return new SourceFragments(fragments);
}

// create two objects:
// - one as dependency with definition
// - second as reference.
// when serializing, if module count only one reference, include it raw.
// when multiple references are encoutered, it will emit `const` declaration

export function identifier(options?: IdentifierOptions): Identifier {
  return new Identifier(options);
}

// export function importId(
//   from: string,
//   name?: string | null,
//   alias?: string | null,
//   noRename?: boolean,
// ): ImportedIdentifier {
//   return new ImportedIdentifier({
//     from,
//     name,
//     alias,
//     noRename,
//   });
// }

/**
 * Emit `const` _name_ `=` _value_ `;`
 *
 * reserve _name_ as identifier
 */
export function constVal(name: string, value: unknown): SourceFragments {
  const id = ts.id(name);
  return ts`const ${id} = ${ts.value(value)};`;
}

/**
 * Emit `export const` _name_ `=` _value_ `;`
 * reserve _name_ as identifier
 */
export function exportVal(name: string, value: unknown): SourceFragments {
  const id = ts.id({ name });
  return ts`export const ${id} = ${ts.value(value)};`;
}

// expose for easy of use
ts.value = value;
ts.id = identifier;
// ts.import = importId;
ts.export = exportVal;
ts.const = constVal;
