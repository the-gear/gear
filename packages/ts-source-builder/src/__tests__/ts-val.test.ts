import { ts } from '../template-tag';

const a = { thisIs: 'A' };
const b = { 'and this is': 'B', with: [1, 2, 3] };
const arr = ['the A:', a, 'and the B', b];
const obj1 = { arr, arr2: arr, arr3: [...arr] };
const obj2 = {
  a: {
    int123: 123,
    obj1,
    nested: { a, b, ref1: obj1 },
  },
};

describe('ts.val', () => {
  it('should serialize simple object', () => {
    expect(ts`${ts.val(a)}`.toString().trim()).toMatchInlineSnapshot(`
      "const gen$ = gen$;
      {
        thisIs: \\"A\\"}"
    `);
  });

  it('should serialize nested objects', () => {
    expect(ts`${ts.val(obj2)}`.toString().trim()).toMatchSnapshot();
  });

  it('should reuse same object used twice', () => {
    const code = ts`
    ${ts.val(a, { name: 'a', isExport: true })}
    [${ts.val(a)}, ${ts.val(a)}]
    `;
    expect(code.toString()).toMatchInlineSnapshot(`
      "const a = a;

      a
      [a, a]
      "
    `);
  });
});
