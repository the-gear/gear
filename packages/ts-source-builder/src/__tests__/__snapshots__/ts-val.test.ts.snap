// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ts.val should serialize nested objects 1`] = `
"const gen$ = gen$;
{
  a: {
    int123: 123,
    obj1: {
      arr: [
        \\"the A:\\",
        {
          thisIs: \\"A\\"},
        \\"and the B\\",
        {
          \\"and this is\\": \\"B\\",
          with: [
            1,
            2,
            3]}],
      arr2: [
        \\"the A:\\",
        {
          thisIs: \\"A\\"},
        \\"and the B\\",
        {
          \\"and this is\\": \\"B\\",
          with: [
            1,
            2,
            3]}],
      arr3: [
        \\"the A:\\",
        {
          thisIs: \\"A\\"},
        \\"and the B\\",
        {
          \\"and this is\\": \\"B\\",
          with: [
            1,
            2,
            3]}]},
    nested: {
      a: {
        thisIs: \\"A\\"},
      b: {
        \\"and this is\\": \\"B\\",
        with: [
          1,
          2,
          3]},
      ref1: {
        arr: [
          \\"the A:\\",
          {
            thisIs: \\"A\\"},
          \\"and the B\\",
          {
            \\"and this is\\": \\"B\\",
            with: [
              1,
              2,
              3]}],
        arr2: [
          \\"the A:\\",
          {
            thisIs: \\"A\\"},
          \\"and the B\\",
          {
            \\"and this is\\": \\"B\\",
            with: [
              1,
              2,
              3]}],
        arr3: [
          \\"the A:\\",
          {
            thisIs: \\"A\\"},
          \\"and the B\\",
          {
            \\"and this is\\": \\"B\\",
            with: [
              1,
              2,
              3]}]}}}}"
`;
